components/browser/icons/src/main/java/mozilla/components/browser/icons/extension/IconMessageHandler.kt

```diff
@@ -0,0 +1,137 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.icons.extension
+
+import androidx.annotation.VisibleForTesting
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import mozilla.components.browser.icons.BrowserIcons
+import mozilla.components.browser.icons.IconRequest
+import mozilla.components.browser.session.Session
+import mozilla.components.concept.engine.EngineSession
+import mozilla.components.concept.engine.webextension.MessageHandler
+import mozilla.components.support.base.log.logger.Logger
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+
+/**
+ * [MessageHandler] implementation that receives messages from the icons web extensions and performs icon loads.
+ */
+internal class IconMessageHandler(
+    private val session: Session,
+    private val icons: BrowserIcons
+) : MessageHandler {
+    private val scope = CoroutineScope(Dispatchers.Main)
+
+    @VisibleForTesting(otherwise = VisibleForTesting.NONE) // This only exists so that we can wait in tests.
+    internal var lastJob: Job? = null
+
+    override fun onMessage(message: Any, source: EngineSession?): Any {
+        if (message is JSONObject) {
```

Comment:
Should this throw an exception on the else case?

Reply:
Good question. Yeah, I think we could do that to uncover unexpected issues.

Reply:
Good question. Yeah, I think we could do that to uncover unexpected issues. I'll update the PR.

Reply:
👍

Reply:
Updated :)